src/browser-installer/registry/index.ts

```diff
@@ -0,0 +1,215 @@
+    };
+
+    browserInstallerDebug(`adding '${name}@${version}' on '${platform}' to registry at ${relativePath}`);
+    outputJSONSync(registryPath, registry, { replacer });
+};
+
+const getBinaryVersions = (name: BinaryName, platform: BrowserPlatform): string[] => {
+    const registryKey = getRegistryKey(name, platform);
+
+    if (!registry[registryKey]) {
+        return [];
+    }
+
+    return Object.keys(registry[registryKey]);
+};
+
+const hasBinaryVersion = (name: BinaryName, platform: BrowserPlatform, version: string): boolean =>
+    getBinaryVersions(name, platform).includes(version);
+
+export const getMatchingDriverVersion = (
+    driverName: SupportedDriver,
+    platform: BrowserPlatform,
+    browserVersion: string,
+): string | null => {
+    const registryKey = getRegistryKey(driverName, platform);
+
+    if (!registry[registryKey]) {
+        return null;
+    }
+
+    if (driverName === Driver.CHROMEDRIVER || driverName === Driver.EDGEDRIVER) {
+        const milestone = getMilestone(browserVersion);
+        const buildIds = getBinaryVersions(driverName, platform);
+        const suitableBuildIds = buildIds.filter(buildId => buildId.startsWith(milestone));
+
+        if (!suitableBuildIds.length) {
+            return null;
+        }
+
+        const suitableBuildIdsSorted = suitableBuildIds.sort(semverVersionsComparator);
+
+        return suitableBuildIdsSorted[suitableBuildIdsSorted.length - 1];
+    }
+
+    if (driverName === Driver.GECKODRIVER) {
+        const buildIds = Object.keys(registry[registryKey]);
+        const buildIdsSorted = buildIds.sort(semverVersionsComparator);
+
+        return buildIdsSorted.length ? buildIdsSorted[buildIdsSorted.length - 1] : null;
+    }
+
+    return null;
+};
+
+export const getMatchingBrowserVersion = (
+    browserName: SupportedBrowser,
+    platform: BrowserPlatform,
+    browserVersion: string,
+): string | null => {
+    const registryKey = getRegistryKey(browserName, platform);
+
+    if (!registry[registryKey]) {
+        return null;
+    }
+
+    let buildPrefix: string;
+
+    switch (browserName) {
+        case Browser.CHROME:
+            buildPrefix = normalizeChromeVersion(browserVersion);
+            break;
+
+        case Browser.CHROMIUM:
+            buildPrefix = getMilestone(browserVersion);
+            break;
+
+        case Browser.FIREFOX:
+            buildPrefix = getFirefoxBuildId(browserVersion);
+            break;
+
+        default:
+            return null;
+    }
+
+    const buildIds = getBinaryVersions(browserName, platform);
+    const suitableBuildIds = buildIds.filter(buildId => buildId.startsWith(buildPrefix));
+
+    if (!suitableBuildIds.length) {
+        return null;
+    }
+
+    const firefoxVersionComparator = (a: string, b: string): number => {
+        a = a.slice(a.indexOf("_") + 1);
+        b = b.slice(b.indexOf("_") + 1);
+
+        return parseInt(a.replace(".", ""), 16) - parseInt(b.replace(".", ""), 16);
```

Comment:
firefox has versions like "stable_131.0a1", "stable_129.0b9" and such. Parsing them without channel prefix ("stable_" or "nightly_") as hex number works

Reply:
> can we solve this problem just with /(\w+)_(\d+)/ regexp?

I don't think so. "129.0b9" is greater than "129.0b7"